### PR TITLE
added TF code for Azure Automation runbook

### PR DESCRIPTION
**Pull request by ericd-mst-github**
opening PR for automation runbook TF code.

### CODE REVIEW

File: quickstart/101-vm-auto-shutdown/main.tf [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    sku       = "2022-datacenter-azure-edition"
+    version   = "latest"
+  }
+
+  boot_diagnostics {
+    storage_account_uri = azurerm_storage_account.my_storage_account.primary_blob_endpoint
+  }
+}
+
+# # Install IIS web server to the virtual machine
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/main.tf]: Why is the following code included if it's commented out?

File: quickstart/101-vm-auto-shutdown/providers.tf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+provider "azurerm" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/providers.tf:12]: Please include the version constraints and required provider blocks. Example: https://github.com/Azure/terraform/blob/master/quickstart/201-k8s-cluster-with-aks-applicationgateway-ingress/providers.tf

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 3]
```diff
@@ -0,0 +1,93 @@
+variable "resource_group_location" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:1]: Please define the resource_group_location and prefix as done in https://github.com/Azure/terraform/blob/master/quickstart/201-k8s-cluster-with-aks-applicationgateway-ingress/variables.tf

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  type        = list(string)
+  default     = ["10.0.1.0/24"]
+}
+
+variable "public_ip_allocation_method" {
+  description = "Allocation method for the public IP."
+  type        = string
+  default     = "Dynamic"
+}
+
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:34]: I believe vm_size has a fixed set of valid values. Please included a validation block. (Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/variables.tf)

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  default     = ["10.0.0.0/16"]
+}
+
+variable "subnet_address_prefixes" {
+  description = "Address prefixes for the subnet."
+  type        = list(string)
+  default     = ["10.0.1.0/24"]
+}
+
+variable "public_ip_allocation_method" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:25]: Please include a validation block for all restricted values. (Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/variables.tf)

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  })
+  default = {
+    publisher = "MicrosoftWindowsServer"
+    offer     = "WindowsServer"
+    sku       = "2022-datacenter-azure-edition"
+    version   = "latest"
+  }
+}
+
+variable "storage_account_tier" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:61]: Please include a validation block for all restricted values. (Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/variables.tf)

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  description = "Performance tier of the storage account."
+  type        = string
+  default     = "Standard"
+}
+
+variable "storage_account_replication_type" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:71]: Please include a validation block for all restricted values. (Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/variables.tf)

File: quickstart/101-vm-auto-shutdown/variables.tf [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  description = "Replication type for the storage account."
+  type        = string
+  default     = "LRS"
+}
+
+variable "automation_account_sku_name" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/variables.tf:81]: Please include a validation block for all restricted values. (Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/variables.tf)

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+## Variables
+
+| Name | Description | Default |
+|-|-|-|
+| `resource_group_location` | The location where the resource group should be created. | East US |
+| `prefix` | A prefix for naming resources. | demo |
+| `vnet_address_space` | Address space for the virtual network. | ["10.0.0.0/16"] |
+| `subnet_address_prefixes` | Address prefixes for the subnet. | ["10.0.1.0/24"] |
+| `public_ip_allocation_method` | Allocation method for the public IP. | Dynamic |
+| `vm_size` | Size of the virtual machine. | Standard_DS1_v2 |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: Needs to list all possible values.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+## Variables
+
+| Name | Description | Default |
+|-|-|-|
+| `resource_group_location` | The location where the resource group should be created. | East US |
+| `prefix` | A prefix for naming resources. | demo |
+| `vnet_address_space` | Address space for the virtual network. | ["10.0.0.0/16"] |
+| `subnet_address_prefixes` | Address prefixes for the subnet. | ["10.0.1.0/24"] |
+| `public_ip_allocation_method` | Allocation method for the public IP. | Dynamic |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: Please update the description to include the possible values for any restricted values. Example: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/readme.md

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+| `vm_size` | Size of the virtual machine. | Standard_DS1_v2 |
+| `vm_image` | Source image reference for the virtual machine. | Windows Server 2022 |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md:35]: Needs to list all possible values.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+| `storage_account_tier` | Performance tier of the storage account. | Standard |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: Needs to list all possible values.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+| `storage_account_replication_type` | Replication type for the storage account. | LRS |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: Needs to list all possible values.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+| `automation_account_sku_name` | SKU name for the Azure Automation Account. | Basic |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: Needs to list all possible values.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
+# Terraform Azure VM Automation Example
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md]: No need to state "Terraform" or "example" as this is the Terraform sample repo. "Azure VM Automation" iis more in keeping with the PG's pattern.

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+| `public_ip_allocation_method` | Allocation method for the public IP. | Dynamic |
+| `vm_size` | Size of the virtual machine. | Standard_DS1_v2 |
+| `vm_image` | Source image reference for the virtual machine. | Windows Server 2022 |
+| `storage_account_tier` | Performance tier of the storage account. | Standard |
+| `storage_account_replication_type` | Replication type for the storage account. | LRS |
+| `automation_account_sku_name` | SKU name for the Azure Automation Account. | Basic |
+| `runbook_type` | Type of the runbook. | PowerShell |
+| `runbook_uri` | URI for the runbook content. | https://raw.githubusercontent.com/azureautomation/runbooks/master/Utility/ASM/Set-AzureScheduleWithRunbook.ps1 |
+| `one_time_schedule_start_time` | Start time for the one-time runbook schedule. | "2023-09-23T00:00:00Z" |
+| `hourly_schedule_start_time` | Start time for the hourly runbook schedule. | "2023-09-23T01:00:00Z" |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md:42]: Needs an "Example" section (https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/readme.md)

File: quickstart/101-vm-auto-shutdown/readme.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+- [azurerm_automation_job_schedule](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/automation_job_schedule)
+- [random_id](https://registry.terraform.io/providers/hashicorp/random/latest/docs/resources/id)
+- [random_password](https://registry.terraform.io/providers/hashicorp/random/latest/docs/resources/password)
+- [random_pet](https://registry.terraform.io/providers/hashicorp/random/latest/docs/resources/pet)
+
+## Variables
+
+| Name | Description | Default |
+|-|-|-|
+| `resource_group_location` | The location where the resource group should be created. | East US |
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/readme.md:29]: For consistency, please see the following readme.md in describing the RG vars: https://github.com/Azure/terraform/blob/master/quickstart/101-front-door-standard-premium/readme.md

File: quickstart/101-vm-auto-shutdown/outputs.tf [hunk 18]
```diff
@@ -0,0 +1,64 @@
+output "resource_group_name" {
```
Comment (TomArcherMsft) [inline quickstart/101-vm-auto-shutdown/outputs.tf:1]: Note: You need to output only the following:
- any value that is dynamically generated (e.g., rg name)
- any value needed to perform a step after the application of the execution plan (so that you can use "terraform output" to output the value to a var to be used with a subsequent azCLI/azPS command).

**Review by TomArcherMsft — CHANGES_REQUESTED** [review]
Great PR, @ericd-mst-github ! Only a few issues to make the sample consistent with the current best practices.

File: quickstart/101-vm-auto-shutdown/main.tf (hunk 1 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/main.tf]: Thanks for catching this. This was modeled from a previous quickstart so the commented code will be removed.

Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/main.tf]: Removed in commit: https://github.com/Azure/terraform/pull/255/commits/f82db3eb77f246a4729dc22a1108d4134d2ceca0

File: quickstart/101-vm-auto-shutdown/providers.tf (hunk 2 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/providers.tf:12]: updated in commit: https://github.com/Azure/terraform/pull/255/commits/efb8d0f25cd62c31e2e9da837b404ca972905f64

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 3 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:1]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 4 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:34]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 5 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:25]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 6 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:61]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 7 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:71]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/variables.tf (hunk 8 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/variables.tf:81]: Thanks for the catch. Updated in commit: https://github.com/Azure/terraform/pull/255/commits/f46699d1f220ccf21bb974cb7faecb545fd08206

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 9 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md]: Updated in commit: https://github.com/Azure/terraform/pull/255/commits/51abcf64bfc1b3480c18885ac0ba87fd998a2bd4

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 10 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md]: Updated in commit: https://github.com/Azure/terraform/pull/255/commits/acabdbdb34c96a8eabb228a2126d285d4f8da3e6

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 13 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md]: Updated in commit: https://github.com/Azure/terraform/pull/255/commits/acabdbdb34c96a8eabb228a2126d285d4f8da3e6

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 15 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md]: Resolved in commit: https://github.com/Azure/terraform/pull/255/commits/51abcf64bfc1b3480c18885ac0ba87fd998a2bd4

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 17 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md:29]: Resolved in commit: https://github.com/Azure/terraform/pull/255/commits/51abcf64bfc1b3480c18885ac0ba87fd998a2bd4

File: quickstart/101-vm-auto-shutdown/outputs.tf (hunk 18 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/outputs.tf:1]: Thanks. I revised the outputs.tf in commit: https://github.com/Azure/terraform/pull/255/commits/e13fcdbffaaae9092854ae80a9088659d05a6e73

File: quickstart/101-vm-auto-shutdown/readme.md (hunk 16 above)
Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md:42]: Thanks again for checking. For the example section, we don't necessarily have a written topic published yet, as that will be a new VM tutorial that covers Auto-shutdown methods (PS, CLI, TF).

Should we have a separate Terraform tutorial doc auto generated through your current method, and I can link to that generated article from the new topic?

Comment (TomArcherMsft) [reply quickstart/101-vm-auto-shutdown/readme.md:42]: @ericd-mst-github 

TL/DR: Please leave the section out and it can be added after the article is published.  

This is a chicken & egg situation where I can't publish the new article until the sample is published. Therefore, I don't know the exact URL until I write the article.  When I'm coding the sample, I generally know what the article name is going to be so I "cheat" by putting in the URL for what will eventually be the published URL. The README includes the invalid URL for only a few hours or a day at most.  However, in this case, you don't know the URL. Therefore, please leave the section out and it can be added after the article is published.

Comment (ericd-mst-github) [reply quickstart/101-vm-auto-shutdown/readme.md:42]: Sounds good @TomArcherMsft and this makes sense to me.

I'll see about leaving the TF section out and go with the CLI/PS methods for now in the VM article. Let me know if further changes are needed for the current TF sample.

**Comment by stemaMSFT** [comment]
LGTM!

**Comment by ericd-mst-github** [comment]
#sign-off
